app/player/handlers.go

```diff
@@ -45,10 +55,41 @@ func (h RequestHandler) processStreamError(w http.ResponseWriter, uri string, er
 	}
 }
 
+func (h *RequestHandler) isViewable(mime string) bool {
+	for _, t := range viewableTypes {
+		if strings.HasPrefix(mime, t) {
+			return true
+		}
+	}
+	return false
+}
+
 // Handle is responsible for all HTTP media delivery via player module.
 func (h *RequestHandler) Handle(w http.ResponseWriter, r *http.Request) {
 	uri := h.getURI(r)
-	err := h.player.Play(uri, w, r)
+	Logger.streamPlaybackRequested(uri, users.GetIPAddressForRequest(r))
+
+	s, err := h.player.ResolveStream(uri)
+	if err != nil {
+		Logger.streamResolveFailed(uri, err)
+		h.processStreamError(w, uri, err)
+		return
+	}
+	Logger.streamResolved(s)
+
+	err = h.player.RetrieveStream(s)
+	if err != nil {
+		Logger.streamRetrievalFailed(uri, err)
+		h.processStreamError(w, uri, err)
+		return
+	}
+	Logger.streamRetrieved(s)
+
+	if !h.isViewable(s.ContentType) || r.URL.Query().Get(ParamDownload) != "" {
+		w.Header().Set("Content-Disposition", fmt.Sprintf("attachment; filename=%v", s.Claim.Value.GetStream().Source.Name))
```

Comment:
just wondering, could this open up to XSS or some sort of attack vector where people add arbitrary code as `name`?

Reply:
This is not executed, it just determines the file name on the user's filesystem.